File: reductus/dataflow/deps.py
# This program is public domain
# Author: Paul Kienzle
"""
Dependency calculator.
"""
from __future__ import print_function

def processing_order(pairs, n=0):
    """
    Order the work in a workflow.

    Given a set of n items to evaluate numbered from zero through n-1,
    and dependency pairs


    :Parameters:

    *pairs* : [(int, int), ...]

        Pairwise dependencies amongst items.

    *n* : int

        Number of items, or 0 if we don't care about any item that is not
        mentioned in the list of pairs

    :Returns:

    *order* : [int, ...]

        Permutation which satisfies the partial order requirements.

    """
    order = _dependencies(pairs)
    if n:
        if any(id >= n for id in order):
            raise ValueError("Not all dependencies are in the set")
        rest = set(range(n)) - set(order)
    else:
        rest = set(k for p in pairs for k in p) - set(order)
    #print "order",order,"from",pairs
    return order + list(rest)


def _dependencies(pairs):
    # type: (List[Tuple[int, int]]) -> List[int]
    emptyset = set()
    order = []  # type: List[int]
    pairs = [(a, b) for a, b in pairs]
    if not pairs:
        return order

    # Break pairs into left set and right set
    left, right = (set(s) for s in zip(*pairs))
    while pairs:
        #print "within",pairs
        # Find which items only occur on the right
        independent = right - left
        if independent == emptyset:
            cycleset = ", ".join(str(s) for s in left)
            raise ValueError("Cyclic dependencies amongst %s" % cycleset)

        # The possibly resolvable items are those that depend on the independents
        dependent = set(a for a, b in pairs if b in independent)
        pairs = [(a, b) for a, b in pairs if b not in independent]
        if not pairs:
            resolved = dependent
        else:
            left, right = (set(s) for s in zip(*pairs))
            resolved = dependent - left
        #print "independent",independent,"dependent",dependent,"resolvable",resolved
        order += resolved
        #print("new order",order)
    order.reverse()
    return order


# ========= Test code ========
def _check(msg, pairs, n):
    """
    Verify that the list n contains the given items, and that the list
    satisfies the partial ordering given by the pairs in partial order.
    """
    order = processing_order(pairs, n=n)
    if len(set(order)) != n:
        raise RuntimeError("%s is missing items" % msg)
    for lo, hi in pairs:
        if order.index(lo) >= order.index(hi):
            raise RuntimeError("%s expect %s before %s in %s for %s"
                               % (msg, lo, hi, order, pairs))
    print("%s %s"%(msg, str(order)))


def test():
    import numpy as np

    # No dependencies
    _check("test empty", [], 9)

    # No chain dependencies
    _check("test2", [(4, 1), (3, 2), (7, 6)], 9)

    # Some chain dependencies
    pairs = [(4, 0), (0, 1), (1, 2), (7, 0), (3, 5)]
    _check("test1", pairs, 9)
    _check("test1 numpy", np.array(pairs), 9)

    # Cycle test
    pairs = [(1, 4), (4, 3), (4, 5), (5, 1)]
    try:
        _ = processing_order(pairs, n=9)
    except ValueError:
        pass
    else:
        raise Exception("test3 expect ValueError exception for %s" % (pairs,))

    # large test for gross speed check
    A = np.random.randint(4000, size=(1000, 2))
    A[:, 1] += 4000  # Avoid cycles
    _check("test-large", A, 8000)

    # depth tests
    k = 200
    A = np.array([range(0, k), range(1, k + 1)]).T
    _check("depth-1", A, 201)

    A = np.array([range(1, k + 1), range(0, k)]).T
    _check("depth-2", A, 201)


if __name__ == "__main__":
    test()

File: reductus/web_gui/webreduce/js/deps.js
const dependencies = {};
export {dependencies};

function order(template, target) {
    var n, pairs, remaining, processed;
    if (target == null) {
        // Order to evaluate all nodes
        pairs = template.wires.map(function(w) {return [w.source[0], w.target[0]]});
        n = template.modules.length;
    }
    return processing_order(pairs, n)
}

function range(n) {
    return Array.from(new Array(n)).map((u,i) => (i))
}

function processing_order(pairs, n) {
    /*
    Order the work in a workflow.

    Given a set of n items to evaluate numbered from zero through n-1,
    and dependency pairs 


    :Parameters:

    *pairs* : [(int, int), ...]

        Pairwise dependencies amongst items.

    *n* : int

        Number of items, or 0 if we don't care about any item that is not
        mentioned in the list of pairs

    :Returns:

    *order* : [int, ...]

        Permutation which satisfies the partial order requirements.

    */
    var n = (n == null) ? 0 : n;
    var order = _dependencies(pairs);
    if (n > 0) {
        var out_of_range = order.filter(function(id) { return id >= n });
        if (out_of_range.length > 0) {
            //if any(id >= n for id in order):
            throw "Not all dependencies are in the set";
        }
        var order_set = new Set(order);
        var rest = new Set(range(n).filter(function(r) { return (!order_set.has(r)) }));
        //set(range(n)) - set(order)
    }
    else {
        var rest = new Set();
        pairs.forEach(function(p) { p.forEach(function(pp) { rest.add(pp) }) });
        //set(k for p in pairs for k in p) - set(order)
    }
    //print "order",order,"from",pairs
    return order.concat(Array.from(rest.values()));
}


function _dependencies(pairs) {
    //print "order_dependencies",pairs
    var emptyset = new Set();
    var order = []

    // Break pairs into left set and right set
    var left = new Set();
    var right = new Set();
    if (pairs.length > 0) {
        left = new Set(pairs.map(function(p) { return p[0] }));
        right = new Set(pairs.map(function(p) { return p[1] }));
    }
    while (pairs.length > 0) {
        // print "within",pairs
        // Find which items only occur on the right
        var independent = new Set();
        right.forEach(function(r) { if (!left.has(r)) { independent.add(r) } }); 
        if (independent.size == 0) {
            var cycleset = left.values().join(", ")
            throw "Cyclic dependencies amongst " + cycleset
        }

        // The possibly resolvable items are those that depend on the independents
        
        var dependent = new Set(pairs.filter(function(p) { return independent.has(p[1]) }).map(function(p) {return p[0]}))
        //set([a for a, b in pairs if b in independent])
        pairs = pairs.filter(function(p) { return (!independent.has(p[1])) });
        //[(a, b) for a, b in pairs if b not in independent]
        if (pairs.length == 0) {
            var resolved = dependent;
        }
        else {
            left = new Set(pairs.map(function(p) { return p[0] }));
            right = new Set(pairs.map(function(p) { return p[1] }));
            //left, right = [set(s) for s in zip(*pairs)]
            var resolved = new Set();
            dependent.forEach(function(d) { if(!left.has(d)) { resolved.add(d) } });
        }
        //print "independent",independent,"dependent",dependent,"resolvable",resolved
        order = order.concat(Array.from(resolved.values()));
    }
        //print "new order",order
    order.reverse()
    return order
}

function mark_satisfied(template, module_defs) {
    var nodes_ordered = order(template);
    var modules_visited = template.modules.map(function(m) { return false; });
    var modules_satisfied = new Set();
    var modules_unsatisfied = new Set();
    var wires_satisfied = new Set();
    var wires_unsatisfied = new Set();
    
    function get_module_satisfied(node) {
        var module = template.modules[node];
        var external_config = (template.config || {})[node];
        var mdef = module_defs[module.module];
        var inputs_satisfied = mdef.inputs.length == 0 || mdef.inputs.every(function(i) { return get_input_satisfied(node, i.id, mdef) });
        var fileinputs = mdef.fields.filter(function(f) { return f.datatype == 'fileinfo' });
        var fileinputs_satisfied = fileinputs.length == 0 || fileinputs.every(function(f) { 
            var embedded = (module.config && module.config[f.id] && module.config[f.id].length > 0);
            var external = (external_config && external_config[f.id] && external_config[f.id].length > 0);
            return embedded || external;
        });
        var satisfaction = (inputs_satisfied && fileinputs_satisfied);
        if (satisfaction) { modules_satisfied.add(node); }
        else { modules_unsatisfied.add(node) }
        return satisfaction;
    }
    
    function get_input_satisfied(node, terminal_id, mdef) {
        var satisfied_in = [];
        var input_def = mdef.inputs.find(function(idef) { return idef.id == terminal_id });
        template.wires.forEach(function(w, i) { 
            if (w.target[0] == node && w.target[1] == terminal_id) {
                if (modules_satisfied.has(w.source[0])) {
                    wires_satisfied.add(i);
                    satisfied_in.push(i);
                } else {
                    wires_unsatisfied.add(i);
                }
            }
        });
        return (!input_def.required) || satisfied_in.length > 0;
    }
    
    nodes_ordered.forEach(function(n) { return get_module_satisfied(n) });
    
    return {
        modules_satisfied, 
        modules_unsatisfied,
        wires_satisfied,
        wires_unsatisfied
    }
}

dependencies.mark_satisfied = mark_satisfied;
